birdhouse/scripts/logging.include.sh

```diff
@@ -0,0 +1,61 @@
+#!/bin/sh
+
+BIRDHOUSE_COLOR=${BIRDHOUSE_COLOR:-1}
+if [ "${BIRDHOUSE_COLOR}" -eq "1" ]; then
+    BLUE=$(tput setaf 12)
+    GRAY=$(tput setaf 8)
+    CYAN=$(tput setaf 6)
+    PURPLE=$(tput setaf 5)
+    YELLOW=$(tput setaf 3)
+    GREEN=$(tput setaf 2)
+    RED=$(tput setaf 1)
+    REG_BG_BOLD="$(tput sgr0)$(tput setab 1)$(tput bold)"
+    NORMAL=$(tput sgr0)
+fi
+
+BIRDHOUSE_LOG_LEVEL=${BIRDHOUSE_LOG_LEVEL:-INFO}
+export LOG_DEBUG="${GRAY}DEBUG${NORMAL}:    "
+export LOG_INFO="${BLUE}INFO${NORMAL}:     "
+export LOG_WARN="${YELLOW}WARNING${NORMAL}:  "
+export LOG_ERROR="${RED}ERROR${NORMAL}:    "
+export LOG_CRITICAL="${REG_BG_BOLD}CRITICAL${NORMAL}: "  # to report misuse of functions
+
+
+# Usage: log {LEVEL} "{message}" [...]
+# Any amount of messages can be passed to the function.
+log() {
+    if [ "${BIRDHOUSE_LOG_LEVEL}" != DEBUG ] \
+    && [ "${BIRDHOUSE_LOG_LEVEL}" != INFO ] \
+    && [ "${BIRDHOUSE_LOG_LEVEL}" != WARN ] \
+    && [ "${BIRDHOUSE_LOG_LEVEL}" != ERROR ]; then
+        echo "${LOG_CRITICAL}Invalid log level setting: [BIRDHOUSE_LOG_LEVEL=${BIRDHOUSE_LOG_LEVEL}]."
+        exit 2
```

Comment:
Existing the entire script just for a typo in the logging level is too agressive for me.

Like for `pavics-compose.sh` I do not fail the entire process if a dir is not found in the `EXTRA_CONF_DIRS` because I do not want to penalize the entire stack (haft `pavics-compose.sh up` process) just for a typo in one of the newly added dir.

Any scripts, even external scripts, can be using this new logging infrastructure.  The script might be doing something very essential and we do not want to break it in the middle of its workflow just for a typo in the logging level.

The worst consequence is simply logging at the wrong level.  I do not think the consequence is severe enough to break the entire script.

Instead of losing the log message, you can simply log it at CRITICAL level.  The person who wants to decrease the verbosity from the default INFO to WARN will quickly realize their mistake without us breaking their script workflow.

Reply:
That's the thing though, nobody actually go through the logs to read them just for fun. The only time we investigate an issue is if something doesn't work. If someone purposely modifies the logging level, at that moment, they would not be running a recurring workflow. I think it is more efficient to fail right away and let the person know to fix it right away, than silently ignoring it and have them find out much later that logs that you needed to investigate a bug are not present because the value configured was invalid.

The only reason I agree that missing dirs from `EXTRA_CONF_DIRS` can be allowed is to support the migration from old configurations to new ones. I don't think the two cases are comparable.

Reply:
> That's the thing though, nobody actually go through the logs to read them just for fun. The only time we investigate an issue is if something doesn't work. 

Ok, then how about treat all log level error as `DEBUG` level?  Then we are sure to not lose any logs?

> If someone purposely modifies the logging level, at that moment, they would not be running a recurring workflow. 

This `BIRDHOUSE_LOG_LEVEL` is set in the `env.local` file.  This file is pretty much part of any recurring workflow.

> find out much later that logs that you needed to investigate a bug are not present because the value configured was invalid.

Totally agree, so I suggested to set all error as `DEBUG` level?  Then no `exit` is needed.

Someone at the commandline interractively will notice it too since it's more verbose than he needs, no need to `exit` to get his attention.

If you wrote some python logging module, would you `sys.exit` inside it?  I think you rather let the consumer of your logging module decide.

Reply:
> If you wrote some python logging module, would you sys.exit inside it? I think you rather let the consumer of your logging module decide.

Just my 2 cents here... I wouldn't `sys.exit` but I would definitely `raise Exception(...)` which would have a similar effect if that exception wasn't handled downstream.

Since shell scripts don't really have a `raise Exception` equivalent then an `exit` achieves the same goal.

Also my opinion... Users will rarely scroll back through a long log output to figure out what is wrong. The best way to highlight that something is wrong to a user is to make sure the _most informative error message_ is the last line in the log/output. The best way to do this is to stop the program when there is an error that users should be immediately aware of.

@tlvu since you're a good developer/sysadmin I'm sure that you're comfortable reading back through long logs to identify sources of errors but this is unusual in my experience (and once more people start using this software we're going to get a lot of questions if the error messages aren't obvious to most users).

Reply:
> once more people start using this software we're going to get a lot of questions if the error messages aren't obvious to most users

This is also my concern.  Given `BIRDHOUSE_LOG_LEVEL` is set in the `env.local` it affects **everything**.  A typo just for a logging level spelling should not break the entire platform.  Remember this is a platform wide log level, not per script or per module log level.

Let's consider this scenario:
* user try out the platform interractively, set log level to `DEBUG` just to debug what is not working
* they found out the cause fixed it
* before going back to prod again, they switch back to `INFO` or `WARN` again but made a typo
* and since it's just logging, they don't bother to run any commands to test. This is expected as logging level should not break anything.
* a few weeks later they realize all their cronjobs scripts did not ever fire since a few weeks because of the typo and the `exit`

Let me re-explain my suggestion, my fault that I wasn't clear explaining the "why":
* do not `exit` but log all the message at `DEBUG`, why? 
  * because in production you want traceability, so having too much logging is less of a problem than losing log info.  In prod, logs are your only friend if you want to investigate what happened in the past.  You can not attach a debugger to something that already happened in the past.  Without enough log, you can not reproduce the problem.
  * in production you want predictability, a typo in a log level is not expected to bring down the entire platform, remember this is not a per script or per module log level, it is a platform wide log level
  * if you are scheduled to go live a feature tomorrow, and you have announced it everywhere, the last thing you want is for autodeploy to not fire and the go-live never happened, just for a log level typo
  * too much logs is a small price too pay to ensure keeping your go-live commitment and traceability in case of problems in prod
* still log the `CRITICAL` message about wrong log level, after the original log message to `DEBUG`, why?
  * as you said, make the error obvious to new users and decrease our support call volume

Reply:
> and since it's just logging, they don't bother to run any commands to test. This is expected as logging level should not break anything.

I think that this is the assumption that we differ on.

>  Given BIRDHOUSE_LOG_LEVEL is set in the env.local it affects everything

But I don't think this is true. 
@fmigneault has not updated the `env.local.example` file to demonstrate that `BIRDHOUSE_LOG_LEVEL` should be set there. Instead the line in `logging.include.sh`:

```
BIRDHOUSE_LOG_LEVEL=${BIRDHOUSE_LOG_LEVEL:-INFO}
```

implies to me that this should be set when the script is called. For example:

```
BIRDHOUSE_LOG_LEVEL=DEBUG ./pavics-compose.sh up -d
```

And this makes sense to me since I don't know if I want the stack deployment to always run at the same log level as a random script.

Reply:
I'm also just going to say. My main priority is to get this PR merged. 

If this is a blocking change for you @tlvu I'm willing to live with the consequences in order to get this PR merged.

Reply:
I share @mishaschwartz's point of view. 

If I was a new user to the platform, trying to set it up for the first time, or even modifying variables later on, I would much rather get an immediate error to know right away how to patch it. The alternative of silently ignoring the error would give me the false impression everything "works", when it doesn't. This gets worse when many such errors accumulate gradually. And to figure out what is the cause of the error basically ask from these new users to go through, and understand, the whole code base, which will lead to a very bad user experience. 

@mishaschwartz interpretation of how logging is configured is correct. The only time an invalid value could happen is the moment that a user overrides it in their own `env.local`. Since they would perform a `pavics-compose ...` call right after, it seems more appropriate to raise the error about misconfigured logging right away, since their immediate intention was to change the logging level. This also takes into consideration that users could see  `BIRDHOUSE_LOG_LEVEL`, and just try `trace` for example, without reading the code. Not failing log calls immediately would lead them to believe this is properly configured.

Reply:
> implies to me that this should be set when the script is called. For example:
> 
> ```
> BIRDHOUSE_LOG_LEVEL=DEBUG ./pavics-compose.sh up -d
> ```
> 
> And this makes sense to me since I don't know if I want the stack deployment to always run at the same log level as a random script.

Agreed.  But since user will be used to the pattern that all configs are saved in `env.local`, it can be there.  In fact, there are many other configs that are not in `env.local.example` but if required, we will have to disclose to users they can also use them in their instantiated `env.local`.

For the individual interactive use-case `BIRDHOUSE_LOG_LEVEL=DEBUG some_scripts` performing an `exit` or log at `CRITICAL` that it is a unknown log level but still run the commands normally, it is the same, user will notice it because it is interactive.

If `BIRDHOUSE_LOG_LEVEL=DEBUG some_scripts` is used in an unattended case, then again, should we surprise the user and break his automation workflow just because of a typo.

I always prefer defensive coding and choose the less bad consequence for production.  In this case, between too much verbose logging vs breaking the entire platform or at least breaking some automation workflow, I'd rather choose verbose logging as a worst case consequence for production.

Reply:
> that it is a unknown log level but still run the commands normally,

Since all the scripts perform sourcing of the `read-configs.incluse.sh`, which itself does a `log` call, this will never happen. The invalid log level will always be flagged immediately before running the script operations.

> should we surprise the user and break his automation workflow

I don't think that would be a surprise. You purposely modify the log level value at that moment (whether from the command line or via `env.local`), and if invalid, it breaks. The surprise would be the other way around, where setting `BIRDHOUSE_LOG_LEVEL=debug` for example seems like valid, but isn't! You have to wait until all the processes are executed, go check your logs, to then notice "why aren't there any logs, I enabled them...". 

> defensive coding and choose the less bad consequence for production

I believe that a better defense is to avoid random values altogether. This is not a personal opinion, it is exactly what all deployment systems, whether docker-compose, terraform, ansible, helm charts, etc. perform. If the configuration does not even respect some kind of "schema validation", it's not even worth trying to deploy the broken config.

Reply:
To summarize, I think our view diverse on this point:

* I value more production use-case with stability, predictability and traceability 

* You value more development use-case with any error failing as fast as possible for new users starting out

We are collaborating.  If you think it is for the best of the platform that we favor development use-case over production use-case, I will submit to the majority.

Reply:
I don't think this discussion is productive anymore. We're arguing about philosophies and assumptions which I'm not sure we're going to agree on.

As I said:

> My main priority is to get this PR merged.

Can we come up with a compromise so that we can move on. This is what I suggest:

Add an additional environment variable named `BIRDHOUSE_FAIL_FAST`, which if true, will exit the program when any critical error message is displayed. If false then it will not exit. 

This way I can set `BIRDHOUSE_FAIL_FAST=1` in my deployment and @tlvu can set `BIRDHOUSE_FAIL_FAST=0` in his deployment and everyone can be happy.

If @fmigneault and @tlvu are ok with this strategy I'm happy to write up a quick PR to implement this.

Reply:
Sounds reasonable to me.

Reply:
> `BIRDHOUSE_LOG_LEVEL=debug` for example seems like valid, but isn't! You have to wait until all the processes are executed, go check your logs, to then notice "why aren't there any logs, I enabled them...".

I guess, my suggested change is still not clear.  Here in code if it helps:

Currently
```sh
echo "${LOG_CRITICAL}Invalid log level setting: [BIRDHOUSE_LOG_LEVEL=${BIRDHOUSE_LOG_LEVEL}]."
exit 2
```
The log message is lost.

My suggestion
```sh
echo "${LOG_DEBUG}$*"
echo "${LOG_CRITICAL}Invalid log level setting: [BIRDHOUSE_LOG_LEVEL=${BIRDHOUSE_LOG_LEVEL}]."
```
We log `DEBUG` the original log message so it is not lost
We display at the end in `CRITICAL` that the log level is wrong, so it is more visible than display it before the original log message.

Same pattern where for the other locations where we `exit` due to wrong log level.

Reply:
> This way I can set `BIRDHOUSE_FAIL_FAST=1` in my deployment and @tlvu can set `BIRDHOUSE_FAIL_FAST=0` in his deployment and everyone can be happy.

What will be the default value for `BIRDHOUSE_FAIL_FAST`?  Are we going to debate about having the `exit` or not vs the default value is `1` or `0`?  It's equivalent.

I believe I made my idea really clear, even with code sample this time.  Like I said I will submit to the majority and I agree with @mishaschwartz we want to move on with this PR and I do not want @mishaschwartz to spend time on yet another config flag that is basically doing the same thing.

Reply:
> What will be the default value for BIRDHOUSE_FAIL_FAST

The default can be 0

Reply:
> > What will be the default value for BIRDHOUSE_FAIL_FAST
> 
> The default can be 0

This means no `exit` with `DEBUG` log of the original log message to not lose it.

Basically back to my suggestion, then why bother with the additional config flag?

If default value of `1` means with `exit`, the current state of the code.

I think we just need to decide between the current state of the code, or my suggestion in comment https://github.com/bird-house/birdhouse-deploy/pull/415#discussion_r1467915752, no need for the extra work to add `BIRDHOUSE_FAIL_FAST`.

Reply:
In order to move on, I suggest we simply leave the code as it is currently. No extra flag or leave how invalid level is handled. I expect that you do not override the default INFO level "just for fun" for your prod env anyway, so you should not have any issue about this setting.

Reply:
> In order to move on, I suggest we simply leave the code as it is currently

:+1: 

-----

Just to clarify a few things:

> Basically back to my suggestion.

I was trying to offer a compromise that you would be ok with. I would of course prefer that the default be `1` but I'm happy to prefer your solution as the default in order to meet you part way on this.

> then why bother with the additional config flag?

To allow for both behaviours, as a compromise so we can both configure the behaviour we want.

Reply:
> > then why bother with the additional config flag?
> 
> To allow for both behaviours, as a compromise so we can both configure the behaviour we want.

Yes I see that, thanks for the suggestion.  I was actually more concerned about how the default behaviors will impact new users.  No matter how many additional config vars we can add, there will only have 1 default value.

For us 3, I think we have discussed about it enough by now that we won't make typo mistake or forget about testing it if we change the logging level.

Let's go with the current state now and see how it goes.